offline HTML doc generation support (#44)
HewlettPackard/hpegl-metal-client: build/html/docs/api_description.md

 
# Introduction

> Welcome to HPE GreenLake for bare metal APIs. 

Author: In GLC user guide. lowercase is used. 
<img width="765" alt="image" src="https://user-images.githubusercontent.com/65302544/199344983-93b990c6-364c-4e92-b408-78d9f12f6d52.png">

'Bare Metal' seems correct to me.  Even Dilip had the same comment. Do you suggest going with that?

> Welcome to HPE GreenLake for bare metal APIs. 

Reviewer: I think it should be capitalized, just like "HPE GreenLake for Private Cloud Enterprise".  I'd say the same for Virtual Machines and Containers as well.  Anyway to get them to change the Table of contents?  Or is that a Marketing decision?  I had asked Bill to make the same change in some other documentation PR recently.

> Welcome to HPE GreenLake for bare metal APIs. 

Author: I know a few folks who had worked on the User Guide. I will reach out to them.

> Welcome to HPE GreenLake for bare metal APIs. 

Reviewer: Official feedback is that BMaaS is not a stand alone solution, so maintaining lower case for 'bare metal'.

The HPE GreenLake for bare metal is included as part of the HPE GreenLake for Private Cloud
Enterprise service, and it is sometimes also called Bare Metal as a Service or BMaaS.

This document describes the HPE GreenLake for bare metal APIs protocol and available endpoints.

The BMaaS API is built on HTTP and is RESTful. It has predictable resource URLs. It retuns HTTP 
response codes to indicate errors. It also accepts and returns JSON in the HTTP body. You can 

> use your favorite HTTP/REST library for your promming language to use HPE GreenLake for Private 

Reviewer 2: promming - programming?

Cloud Enterprise bare metal APIs.

The APIs provides access to bare metal service within a single project context. Clients can create 

> a fully-provisioned hosts, storage volumes, and project-specific private networks in an isolated 

Reviewer 2: ```suggestion
fully-provisioned hosts, storage volumes, and project-specific private networks in an isolated 
```

> project environment.  Clients can also access information about available services and resources 

Reviewer 3: ```suggestion
project environment. Clients can also access information about available services and resources 
```

> through the available-resources API.  This API provides detailed information about the OS imaging 

Reviewer 3: ```suggestion
through the available-resources API. This API provides detailed information about the OS imaging 
```

> options, the machine size options, the storage volume options, data center locations,and such that 

Reviewer 3: ```suggestion
options, the machine size options, the storage volume options, data center locations and such that 
```

> options, the machine size options, the storage volume options, data center locations,and such that 

Reviewer 3: ```suggestion
options, machine size options, storage volume options, data center locations which
```

are needed when creating hosts and volumes.

> The API specification references a few resources by the old name, which differs from the name 

Reviewer 2: Since this will be the first version provided externally, should we be using something other than "old" to explain the mapping?

Author: does 'obsolete' fit better?

Reviewer 2: Yes -- obsolete should be good to use.

Author: used 'deprecated' instead of 'old'/'obsolete'.

displayed in HPE GreenLake Central UI. The table below maps the resource's old name in the API 
specification with the corresponding name in HPE GreenLake Central UI.

| **SN** | **Name in API Spec** | **Name in GreenLake Central** |
|--------|----------------------|-------------------------------|
| 1      | Project              | Compute Group                 |
| 2      | Host                 | Compute Instance              |
| 3      | Machine Size         | Compute Instance Type         |

# Authentication

## Bearer Auth

The HPE GreenLake for bare metal APIs uses Bearer Authentication that requires the users to provide 
a bearer token in the **Authorization** header.   The service supports two types of tokens, the HPE 

> GreenLake IAM token and the Metal classic token.  The Metal classic token is HPE internal purpose only.

Reviewer: ```suggestion
GreenLake IAM token and the Metal classic token.  The Metal classic token is for HPE internal use only.
```

> GreenLake IAM token and the Metal classic token.  The Metal classic token is HPE internal purpose only.

Reviewer 3: ```suggestion
GreenLake IAM token and the Metal classic token. The Metal classic token is HPE internal purpose only.
```

The clients are required to provide HPE GreenLake IAM token to authenticate with the HPE GreenLake 

> for bare metal APIs.  The access token can be obtained using any of the following way.

Reviewer 2: ```suggestion
for bare metal APIs.  The access token can be obtained using any of the following ways.
```

<h2>Option1: API client(recommended)</h2>

An API client allows nonhuman entities (an application service account, for instance) programmatic access to a resource on a space.

  + 1.
<a
    href="https://support.hpe.com/hpesc/public/docDisplay?docId=a00092451en_us&page=greenlakecentral-create-api-client.html"
    target="_blank">
    Create an API client
    <img src="https://raw.githubusercontent.com/grommet/grommet-icons/stable/img/share.svg"
    height="12">
</a>

**Note:** Make sure to save the **Issuer Url**, **Client ID**, and **Client Secret**

  + 2.
<a
    href="https://support.hpe.com/hpesc/public/docDisplay?docId=a00092451en_us&page=GUID-1CEA233B-C4B0-41B7-9A25-7A36D9FC0312.html"
    target="_blank">
    Creating an assignment for an API client
    <img src="https://raw.githubusercontent.com/grommet/grommet-icons/stable/img/share.svg"
    height="12">
</a>

> **Note:** Make sure to assign the appropriate  HPE GreenLake for bare metal user roles in HPE GreenLake Central.

Reviewer: ```suggestion
**Note:** Make sure to assign the appropriate HPE GreenLake for Bare Metal user roles in HPE GreenLake Central.
```

  + 3. Make a REST call to generate the API access token:

```
curl -i -X POST \
  '{issuerUrl}/v1/token' \
  -H 'Content-Type: application/x-www-form-urlencoded' \
  -d 'client_id={clientID}' \
  -d 'client_secret={clientSecret}' \
  -d grant_type=client_credentials \
  -d scope=hpe-tenant
```
Obtain the `access_token` from the response.

> <h2>Option2: Getting the token directly from UI</h2>

Reviewer: ```suggestion
<h2>Option 2: Getting the token directly from UI</h2>
```

To authenticate with the BMaaS API you need to obtain the access token from
<a
    href="https://client.greenlake.hpe.com"
    target="_blank">
    HPE GreenLake Central
    <img src="https://raw.githubusercontent.com/grommet/grommet-icons/stable/img/share.svg"
    height="12">
</a>.
Once logged in to **HPE GreenLake Central**,
Click the
<img src="https://raw.githubusercontent.com/grommet/grommet-icons/stable/img/user.svg">
icon in the top right corner and then click
<input
    type="button"
    value="API Access"
    style="color: #000;
        background-color: #fff;
        border: 2px solid #00b388;
        border-radius: 4px;"/>
. In the API Access page, click the
<img src="https://raw.githubusercontent.com/grommet/grommet-icons/stable/img/clipboard.svg">
icon to copy the personal access token.

<br>

> > <img src="https://raw.githubusercontent.com/grommet/grommet-icons/stable/img/circle-information.svg"> **INFO**

Reviewer: Is there a mismatch of bracket here?  The leading ">" doesn't seem to have a match.

Author: that was a typo. removed it.

> > <img src="https://raw.githubusercontent.com/grommet/grommet-icons/stable/img/circle-information.svg"> **INFO**

Reviewer 2: Is the leading '>' intentional?

Author: it was a typo. fixed now, thanks,

For both the options, the access token lease expires in **15** minutes.

## Project API Key

> The specific Project ID must be provided with **Project** header required when accessing service

Reviewer 2: ```suggestion
The specific Project ID must be provided with **Project** header when accessing service
```

within a single project context. Used only when using HPE GreenLake IAM issued access tokens.

## Membership API Key
HPE Internal only. Required only when using classic Metal access tokens.

# Getting a Space name or ID

> Space name or ID is required in header for few Project level APIs. You can obtain the Space name and the corresponding ID from 

Reviewer: ```suggestion
Space name or ID is required in header for a few Project level APIs. You can obtain the Space name and the corresponding ID from 
```

<a
    href="https://client.greenlake.hpe.com"
    target="_blank">
    HPE GreenLake Central
    <img src="https://raw.githubusercontent.com/grommet/grommet-icons/stable/img/share.svg"
    height="12">
</a>.   
Once logged in to **HPE GreenLake Central**,
Click the
<img src="https://raw.githubusercontent.com/grommet/grommet-icons/stable/img/user.svg">
icon in the top right corner and copy the value of **Space** field.  

The corresponding ID for the Space can be extracted from the current URL in the address bar.  
It is the value of **spaceid** query parameter at the end of a URL after a '?' symbol.

> # HPE GreenLake for bare metal API

Reviewer 2: Does this go to the top of file?

Author: mike had the same comment. closing this and keeping the other conversation open.

> # HPE GreenLake for bare metal API

Author: @mchuang3 @anurag-shrivastav  It was intentionally kept to create logical grouping as in this picture.
<img width="882" alt="image" src="https://user-images.githubusercontent.com/65302544/199344485-63fc6258-a3ff-45de-981b-fe10d9f4e930.png">
